wechatpay/src/main/java/com/stripe/android/payments/wechatpay/WeChatPayAuthActivity.kt

```diff
@@ -3,41 +3,125 @@ package com.stripe.android.payments.wechatpay
 import android.app.Activity
 import android.content.Intent
 import android.os.Bundle
+import android.util.Log
+import androidx.annotation.VisibleForTesting
 import androidx.appcompat.app.AppCompatActivity
 import com.stripe.android.StripeIntentResult
 import com.stripe.android.exception.StripeException
 import com.stripe.android.model.WeChat
 import com.stripe.android.payments.PaymentFlowResult
+import com.stripe.android.payments.wechatpay.reflection.DefaultWeChatPayReflectionHelper
+import com.stripe.android.payments.wechatpay.reflection.WeChatPayReflectionHelper
+import java.lang.reflect.InvocationHandler
+import java.lang.reflect.Method
+
+internal class WeChatPayAuthActivity : AppCompatActivity(), InvocationHandler {
+    // TODO(ccen): inject reflectionHelper as a singleton
+    @VisibleForTesting
+    internal var reflectionHelperProvider: () -> WeChatPayReflectionHelper =
+        { DefaultWeChatPayReflectionHelper() }
+    private val reflectionHelper by lazy { reflectionHelperProvider() }
+    private val weChatApi by lazy { reflectionHelper.createWXAPI(this) }
+    private val iWXAPIEventHandler by lazy { reflectionHelper.createIWXAPIEventHandler(this) }
+    private var clientSecret: String? = null
 
-internal class WeChatPayAuthActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         runCatching {
```

Comment:
Did you test this logic with "Don't keep activities" enabled?

Reply:
Just verified with this option on, the same `WeChatPayAuthActivity`'s `#onNewIntent` is correctly called when redirected back from wechat app.

Reply:
I'm worried of a possible race condition between `launchWeChat()` being called again through `onCreate()` and `onNewIntent()`